calculateMu.R

```diff
@@ -22,60 +22,59 @@ calculate.mu <- function(data, input, od_select) {
+      ## input:
+      # time_x, vector: measured time values
+      # od_values_y, vector: accompanying OD values
+      # max_OD_value, default=0.2, float: local OD maxima should be above this value
+      # filter_threshold, default=TRUE, bool: should max_OD_value be applied or not?
+      ## output:
+      # data frame with the following columns: max.pos, max.x, max.y, mu, r.squared, length.iv, residuals
+      
+      ## First step: identify local maxima
+      # identify positions at which dilution took place (OD value drops), accept some downwards fluctuations in OD measurements
+      positions_before_drop <- diff(od_values_y) < (-0.01) # vector of boolean
+      # filter out stretches where dilution took place over several steps, i.e. stretches of "TRUE" - keep first in stretch
+      indices_local_maxima_prelim <- which(positions_before_drop)
+      differences_indices <- diff(indices_local_maxima_prelim)
+      for(i in 1:length(differences_indices)){
+        if(differences_indices[i]==1){
+          positions_before_drop[indices_local_maxima_prelim[i]+1] <- FALSE
         }
-        # filter local maxima by manual threshold
-        y[x[y] > UserMaxInterval]
       }
-      
-      # positions of local maxima
-      max.pos <- localMaxima(y)
-      
-      # positions of local minimum in intervals
-      if (UserMinSelect %in% c("auto", "min-max")) {
-        # compute interval between two local maxima
-        intervals <-
-          mapply(seq, c(1, max.pos[-length(max.pos)] + 1), max.pos)
-        # determine minimum as local minimum of the interval
-        # to correct for lag phases
-        min.pos <- unlist(lapply(intervals, function(iv)
-          iv[which.min(y[iv])])) 
-      } else {
-        # or alternatively, minimum is n steps backward from max, user defined
-        max.pos <- max.pos[max.pos > as.numeric(UserMinSelect)]
-        min.pos <- max.pos - as.numeric(UserMinSelect)
+      indices_local_maxima <- which(positions_before_drop)
+      if(filter_threshold){
+        indices_local_maxima <- indices_local_maxima[od_values_y[indices_local_maxima]>max_OD_value]
       }
       
-      # determine growth rate in intervals from local minima to maxima
-      mapply(function(min.pos, max.pos) {
-        # in most simple case, only min and max are considered
-        if (UserMinSelect == "min-max") {
-          index <- c(min.pos, max.pos)} 
-        else {
-          index <- seq(min.pos, max.pos)
-        }
-        index <- index[y[index] > 0]
-        model <- lm(log(y) ~ x, data.frame(x = x[index], y = y[index]))
-        list(
-          max.pos = max.pos,
-          max.x = x[max.pos],
-          max.y = y[max.pos],
-          mu = coefficients(model)[[2]],
-          r.squared = summary(model)$r.squared,
-          length.iv = length(index),
-          residuals = sd(model$residuals)
-        )
-      }, min.pos, max.pos) %>%
-      t %>% apply(2, unlist) %>% as.data.frame
+      ## Second step: identify minima
```

Comment:
Thanks for this apparently nice solution. I tested it and it seems to work. Is the outcome the same as before?
Also important: can this code be re-intended / formatted? Usually code should not exceed 80 chars, but some of these lines are extremely long like 250 chars. I'd suggest RStudio --> Code --> reformat code and break up comments on multiple lines

Reply:
Formatting: Will take care of this

Outcome: The code gives, in some cases, different results. I worked on it because the previous code sometimes gave quite un-intuitive growth rates and the code I am suggesting is closer to what I would expect. 

See example with screenshots from the app, this example shows four channels in turbidostat mode, would assume from sawtooth pattern that growth rates should be high in the beginning and end of cultivation, lower between: 
![image](https://github.com/m-jahn/ShinyMC/assets/83819421/608b0c93-d03e-4102-89e4-2cbbaa077cb5)

These are the growth rates analysed with the old code: 
![image](https://github.com/m-jahn/ShinyMC/assets/83819421/3538f407-b5ff-40f2-a311-ae4281756081)

And with the new code:
![image](https://github.com/m-jahn/ShinyMC/assets/83819421/192b3e9a-e717-41cf-93ca-02c75ad56443)

Compare also when testing the code outside of ShinyMC the identification of local maxima (coloured red) and minima (coloured green) with the new code: 
![image](https://github.com/m-jahn/ShinyMC/assets/83819421/ddde2b3c-3945-448d-9688-2a6299a12822)

and the old code: 
![image](https://github.com/m-jahn/ShinyMC/assets/83819421/ddc30389-55e9-4a2a-a3eb-8c4e9609de57)

Reply:
Looks much better :-)